### DISCUSSION TITLE
Validating FybrikModules values during run-time without including restrictions in the CRD

### DISCUSSION THREAD

**Discussion by ronenkat** [fybrik/fybrik#776] [Developers] [answered]
In Fybrik we have chosen to implement a registry (catalog) of data path modules as a Kubernetes CR, hence `FybrikModule` represent a module that can be install during runtime.

The goal of the taxonomy work is to allow updating the set of valid values for a `FybrikModule` without changing the `FybrikModule` CRD. 
We will be removing 
With the taxonomy work part of the `FybrikModule` yaml will not be check as part of the custom resource validations in order to make part of the yaml dynamic --- specifically allowing adding/removing valid values without re-creating the CRDs.

The `FybrikModule` doesn't have a controller associated with it.

Options for validating the content of the `FybrikModule` CR:
1. Webhook - (note that webhooks are not enabled in some environment due to operational concerns)
2. Reconciler - require adding a controller for `FybrikModule` 
3. Deploy a `FybrikModule` CR only through an "installer" tool - i.e., create an option for Fybrik command line to install `FybrikModule` CRs - requires using a command line tool
4. Verify during runtime by the plotter control that the `FybrikModule` "entry" in the modules catalog is valid.

Thoughts? opinions?

**Comment by roee88** [comment]
I think that adding a controller is fine, with indication in the status that it was registered correctly. @froesef also suggested indication of applicability of a module to specific remote clusters (e.g., when the module requires a cluster wide dependency).

**Reply by shlomitk1** [reply]
Agree with @roee88 
Option 3 will not allow checking of modules that are manually installed during development/testing by people contributing modules or anyone who evaluates Fybrik.
Option 4 will affect performance.
In the future we may want to delegate some of the orchestration tasks to the module controller, e.g. orchestrate dependencies and check their state. Currently, manager does some of the work (checks that dependent FybrikModules are deployed).

**Reply by simanadler** [reply]
> The goal of the taxonomy work is to allow updating the set of valid values for a `FybrikModule` without changing the `FybrikModule` CRD.

I think that the taxonomy work as it relates to FybrikModules is twofold:
1) Verify that the values in FybrikModules are aligned with the taxonomy - i.e. they are "legal" values - which you have addressed 

2) FybrikModules define the capabilities that are available to be included in data paths.  
a) When creating a new capability, there needs to be a way to update the taxonomy with the new capability, action, etc.  
b) Once included in the taxonomy, these then become available for use in policies.  Do we need to indicate an update to the policy manager?  How does it know there are new capabilities?

==> How do you plan to address this?

> Options for validating the content of the `FybrikModule` CR:
> 
>     1. Webhook - (note that webhooks are not enabled in some environment due to operational concerns)
> 
>     2. Reconciler - require adding a controller for `FybrikModule`
> 
>     3. Deploy a `FybrikModule` CR only through an "installer" tool - i.e., create an option for Fybrik command line to install `FybrikModule` CRs - requires using a command line tool
> 
>     4. Verify during runtime by the plotter control that the `FybrikModule` "entry" in the modules catalog is valid.
> 
> 
> Thoughts? opinions?

I agree validation should probably be done via a FybrikModule controller.  However, how will you differentiate between an "invalid value" vs the introduction of a new capability, action, etc?

**Reply by roee88** [reply]
@simanadler we are not yet addressing this. We* decided that in the _short term_ - a module doesn't bring its own capability but rather refers to an existing taxonomy (same as the approach where actions are owned by policy manager). The longer term approach needs a more careful design and I'm not sure if it's in scope for this discussion. The technical solution of validation doesn't need to change between the two (because you just add an extra pre step to update the taxonomy).

\* "We" includes you :)

**Comment by ronenkat** [comment, marked as answer]
Based on the discussion above the choice for validating the FybrikModule values should be
* Reconciler - require adding a controller for FybrikModule

**Reply by simanadler** [reply]
@ronenkat Do you want to open an issue for this work?

**Reply by ronenkat** [reply]
@simanadler FYI - the work is already tracked in #613
